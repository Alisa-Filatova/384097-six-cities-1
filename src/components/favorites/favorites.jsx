import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import CityTab from '../city-tab/city-tab.jsx';
import OfferCard from '../offer-card/offer-card.jsx';
import AppFooter from '../app-footer/app-footer.jsx';
import {getFavoriteOffers} from '../../reducers/data/selectors';

const Favorites = ({offers}) => {
  return (
    <>
      {offers.length > 0 ?
        <main className="page__main page__main--favorites">
          <div className="page__favorites-container container">
            <section className="favorites">
              <h1 className="favorites__title">Saved listing</h1>
              <div className="favorites__list">
                {offers.map((offer) =>
                  <div
                    className="favorites__locations-items"
                    key={offer.city.name + offer.id}
                  >
                    <div className="favorites__locations locations locations--current">
                      <CityTab city={offer.city} isActive />
                    </div>
                    <div className="favorites__places">
                      <OfferCard
                        offer={offer}
                        prefix="favorites"
                        small
                      />
                    </div>
                  </div>
                )}
              </div>
            </section>
          </div>
        </main>
        :
        <main className="page__main page__main--favorites page__main--favorites-empty">
          <div className="page__favorites-container container">
            <section className="favorites favorites--empty">
              <h1 className="visually-hidden">Favorites (empty)</h1>
              <div className="favorites__status-wrapper">
                <b className="favorites__status">Nothing yet saved.</b>
                <p className="favorites__status-description">
                  Save properties to narrow down search or plan your
                  future trips.
                </p>
              </div>
            </section>
          </div>
        </main>
      }
      <AppFooter />
    </>
  );
};

Favorites.propTypes = {
  offers: PropTypes.arrayOf(PropTypes.object),
};

const mapStateToProps = (state, ownProps) => (
  Object.assign({}, ownProps, {
    offers: getFavoriteOffers(state),
  })
);

export {Favorites};
export default connect(mapStateToProps)(Favorites);


