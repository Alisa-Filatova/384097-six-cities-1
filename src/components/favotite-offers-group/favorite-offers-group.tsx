import * as React from 'react';
import OfferCard from '../offer-card/offer-card';
import CityTab from '../city-tab/city-tab';
import {CityOffersGroup, Offer} from '../../types/offer';

interface Props {
  offersGroup: CityOffersGroup;
  onFavoriteClick: (offer: Offer) => void;
  isAuthenticated: boolean;
}

const FavoriteOffersGroup = (props: Props) => {
  const {offersGroup, onFavoriteClick, isAuthenticated} = props;

  return (
    <div className="favorites__list">
      {Object.entries(offersGroup).map(([key, value]) => (
        <div
          className="favorites__locations-items"
          key={key}
        >
          <div className="favorites__locations locations locations--current">
            <CityTab city={key} isActive />
          </div>
          <div className="favorites__places">
            {Object.values(value).map((offer, idx) => (
              <OfferCard
                offer={offer}
                onFavoriteClick={onFavoriteClick}
                isAuthenticated={isAuthenticated}
                key={`${offer.title}${idx}`}
                prefix="favorites"
                small
              />
            ))}
          </div>
        </div>
      ))}
    </div>
  );
};

export default FavoriteOffersGroup;
