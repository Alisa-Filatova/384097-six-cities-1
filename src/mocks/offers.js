export default [
  {
    id: 123,
    title: `Beautiful & luxurious apartment at great location`,
    img: `img/apartment-01.jpg`,
    isPremium: true,
    price: 120,
    stars: 4,
    type: `Apartment`,
    isInBookmarks: false,
  },
  {
    id: 234,
    title: `Wood and stone place`,
    img: `img/apartment-02.jpg`,
    isPremium: false,
    price: 80,
    stars: 2,
    type: `Private room`,
    isInBookmarks: true,
  },
  {
    id: 456,
    title: `Canal View Prinsengracht`,
    img: `img/apartment-03.jpg`,
    isPremium: false,
    price: 180,
    stars: 3,
    type: `Apartment`,
    isInBookmarks: false,
  },
  {
    id: 567,
    title: `Nice, cozy, warm big bed apartment`,
    img: `img/apartment-01.jpg`,
    isPremium: false,
    price: 100,
    stars: 4,
    type: `Private room`,
    isInBookmarks: true,
  },
];
