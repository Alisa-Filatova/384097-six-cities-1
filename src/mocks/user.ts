export const userMock = {
  avatarUrl: `/static/avatar/4.jpg`,
  email: `alicefill@y888.ru`,
  id: 1,
  isPro: false,
  name: `alicefill`,
};
