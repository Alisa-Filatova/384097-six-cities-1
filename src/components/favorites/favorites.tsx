import * as React from 'react';
import {connect} from 'react-redux';
import AppFooter from '../app-footer/app-footer';
import FavoriteOffersGroup from '../favotite-offers-group/favorite-offers-group';
import withRedirectRoute from '../../hocs/with-redirect-route/with-redirect-route';
import {getFavoriteOffersGroupedByCity} from '../../reducers/data/selectors';
import {Operation as DataOperation} from '../../reducers/data/data';
import {getAuthorizationStatus} from '../../reducers/user/selectors';
import {CityOffersGroup, Offer} from '../../types/offer';
import {ROUTES} from '../../constants/constants';

interface Props {
  offers: CityOffersGroup;
  onFavoriteClick: (offer: Offer) => void;
  isAuthenticated: boolean;
  loadFavoriteOffers: () => void;
}

class Favorites extends React.PureComponent<Props> {

  componentDidMount() {
    this.props.loadFavoriteOffers();
  }

  render() {
    const {offers, isAuthenticated, onFavoriteClick} = this.props;

    return (
      <>
        {Object.keys(offers).length > 0 ? (
          <main className="page__main page__main--favorites">
            <div className="page__favorites-container container">
              <section className="favorites">
                <h1 className="favorites__title">Saved listing</h1>
                <FavoriteOffersGroup
                  offersGroup={offers}
                  isAuthenticated={isAuthenticated}
                  onFavoriteClick={onFavoriteClick}
                />
              </section>
            </div>
          </main>
        ) : (
          <main
            className="page__main page__main--favorites page__main--favorites-empty"
            style={{height: `70vh`}}
          >
            <div className="page__favorites-container container">
              <section className="favorites favorites--empty">
                <h1 className="visually-hidden">Favorites (empty)</h1>
                <div className="favorites__status-wrapper">
                  <b className="favorites__status">Nothing yet saved.</b>
                  <p className="favorites__status-description">
                    Save properties to narrow down search or plan your
                    future trips.
                  </p>
                </div>
              </section>
            </div>
          </main>
        )}
        <AppFooter/>
      </>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  ...ownProps,
  offers: getFavoriteOffersGroupedByCity(state),
  isAuthenticated: getAuthorizationStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onFavoriteClick: (offer) => {
    dispatch(DataOperation.toggleFavorite(offer));
  },
  loadFavoriteOffers: () => {
    dispatch(DataOperation.loadFavoriteOffers());
  }
});

export {Favorites};
export default withRedirectRoute(
    connect(mapStateToProps, mapDispatchToProps)(Favorites),
    ROUTES.LOGIN,
    false
);
