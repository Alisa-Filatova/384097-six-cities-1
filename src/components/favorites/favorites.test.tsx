import * as React from 'react';
import * as Adapter from 'enzyme-adapter-react-16';
import {configure, shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import {Favorites} from './favorites';
import {favoriteOffersGroupMock} from '../../mocks/favorite-offers-group';

configure({adapter: new Adapter()});

describe('Favorites', () => {
  it('renders correctly', () => {
    const page = shallow(
        <Favorites
          offers={favoriteOffersGroupMock as any}
          isAuthenticated={true}
          onFavoriteClick={jest.fn()}
          loadFavoriteOffers={jest.fn()}
        />
    );

    expect(shallowToJson(page)).toMatchSnapshot();
  });
});
