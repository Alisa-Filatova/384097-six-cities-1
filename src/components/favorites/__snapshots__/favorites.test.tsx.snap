// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites renders correctly 1`] = `
<Fragment>
  <main
    className="page__main page__main--favorites"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <FavoriteOffersGroup
          isAuthenticated={true}
          offersGroup={
            Object {
              "Amsterdam": Array [
                Object {
                  "bedrooms": 3,
                  "city": Object {
                    "location": Object {
                      "latitude": 52.370216,
                      "longitude": 4.895168,
                      "zoom": 10,
                    },
                    "name": "Amsterdam",
                  },
                  "description": "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.",
                  "goods": Array [
                    "Heating",
                    "Kitchen",
                    "Cable TV",
                    "Washing machine",
                    "Coffee machine",
                    "Dishwasher",
                  ],
                  "host": Object {
                    "avatarUrl": "img/1.png",
                    "id": 3,
                    "isPro": true,
                    "name": "Angelina",
                  },
                  "id": 1,
                  "images": Array [
                    "img/1.png",
                    "img/2.png",
                  ],
                  "isFavorite": true,
                  "isPremium": false,
                  "location": Object {
                    "latitude": 52.35514938496378,
                    "longitude": 4.673877537499948,
                    "zoom": 8,
                  },
                  "maxAdults": 4,
                  "previewImage": "img/1.png",
                  "price": 120,
                  "rating": 4.8,
                  "title": "Beautiful & luxurious studio at great location",
                  "type": "apartment",
                },
              ],
            }
          }
          onFavoriteClick={[MockFunction]}
        />
      </section>
    </div>
  </main>
  <AppFooter />
</Fragment>
`;
